Local feeds: Unit tests for LocalFeedUpdater

damoasda opened the pull request:
Provides unit tests for the class `LocalFeedUpdater`.

Uses [Robolectric](http://robolectric.org/) and Mockito to run the tests without an Android emulator to make it faster.
Does that way of testing fit to the AntennaPod code style? I'm happy to get feedback.

Contributes to #4287.

ByteHamster commented on core/src/test/java/de/danoeh/antennapod/core/feed/LocalFeedUpdaterTest.java:
```diff
@@ -0,0 +1,258 @@
+        ClientConfig.applicationCallbacks = mock(ApplicationCallbacks.class);
+        when(ClientConfig.applicationCallbacks.getApplicationInstance()).thenReturn(app);
+
+        localFeedDir1 = new File("sampledata/local-feed1");
```
Would it be possible to use a test flavour of the `assets` folder? From my understanding, `sampledata` is used for layout rendering in Android Studio, not for unit tests.

damoasda replied:
Thank you for the hint regarding `sampledata`, I wasn't aware of that. I will look for an alternative solution in the next days.

damoasda replied:
I moved the local feed media files from the `sampledata` folder to the `test/assets` folder. For file access `context.getAssets()` is used. Anything else I can improve?

ByteHamster commented on core/src/test/java/de/danoeh/antennapod/core/feed/LocalFeedUpdaterTest.java:
```diff
@@ -0,0 +1,258 @@
+        adapter.open();
+        adapter.close();
+
+        // Emulate turned off gpodder.net support in SyncService
```
Does this need to be emulated manually? I think the default value of gpodder should already be set to "off".

damoasda replied:
The comment is misleading somehow. The core point is that `ClientConfig.applicationCallbacks` has to be initialized manually to make `GpodnetPreferences.loggedIn()` return false, otherwise I get a `NullPointerException` in `GpodnetPreferences`.

ByteHamster replied:
Ah okay. How about calling `ClientConfig.initialize`? That should deal with initializing PodDbAdapter, user preferences, etc.

damoasda replied:
Hmm, `ClientConfig.initialize()` doesn't set `ClientConfig.applicationCallbacks`, so I get a `NullPointerException` in `GpodnetPreferences`. The only code that currently sets `ClientConfig.applicationCallbacks` is in the class `ClientConfigurator` that belongs to the `app` module. Did I miss something?

ByteHamster replied:
Oh my, those callback methods regularly annoy me :D They are a leftover from the time when AntennaPod could be easily white-labelled.

Would it be possible to only mock `applicationCallbacks` but not shared preferences?

damoasda replied:
Yes, I could simplify the test code. Now it only mocks the class `ApplicationCallbacks` but not the classes `Application` and `SharedPreferences`.

ByteHamster commented on core/src/test/java/de/danoeh/antennapod/core/feed/LocalFeedUpdaterTest.java:
```diff
@@ -0,0 +1,258 @@
+    @SuppressWarnings("JavaReflectionMemberAccess")
+    @After
+    public void tearDown() throws Exception {
+        // Workaround for Robolectric issue in ShadowSQLiteConnection: IllegalStateException: Illegal connection pointer
```
This looks pretty hacky. What about adding a method directly to `PodDbAdapter`? Something like `PodDbAdapter.tearDownTests()`?

damoasda replied:
You're right, it's ugly code. :sob:
An approach like `PodDbAdapter.tearDownTests()` couldn't cover all needs without the help of Mockito as the field `SQLiteOpenHelper.mContext` is declared as final, but it has to be set to null after each test to avoid the "Illegal connection pointer" error when using Robolectric. For more information see the discussion in https://github.com/robolectric/robolectric/issues/1890.

ByteHamster replied:
Hmm, okay then I would suggest to at least hide the ugliness somewhere in another class (basically like `EspressoTestUtils`, just for `RobolectricTestUtils`)

damoasda replied:
Sorry, my answer is wrong. The field `SQLiteOpenHelper.mContext` is not touched, instead the field `SQLiteOpenHelper.mDatabase` has to be reset which is not final. Finally, I could introduce the method `PodDbAdapter.tearDownTests()` as suggested and replace the hacky reflection code. :grin:

ByteHamster commented on core/src/test/java/de/danoeh/antennapod/core/feed/LocalFeedUpdaterTest.java:
```diff
@@ -0,0 +1,258 @@
+            dfMock.when(() -> DocumentFile.fromTreeUri(any(), any())).thenReturn(documentFile1);
+            Feed feed = new Feed(FEED_URL, null);
+            LocalFeedUpdater.updateFeed(feed, context);
+        }
```
Adding items seems to be pretty repetitive code. How about creating a function for that?

damoasda replied:
Do you mean to extract a helper method with just the following two lines?

```
Feed feed = new Feed(FEED_URL, null);
LocalFeedUpdater.updateFeed(feed, context);
```

ByteHamster replied:
I meant the whole thing about mocking external storage and updating the feed. That looks pretty similar for all tests (I have only skimmed the code, though. Maybe I have overlooked something).

```
	DocumentFile documentFile = DocumentFile.fromFile(localFeedDir2);
        try (MockedStatic<DocumentFile> dfMock = Mockito.mockStatic(DocumentFile.class)) {
            // mock external storage
            dfMock.when(() -> DocumentFile.fromTreeUri(any(), any())).thenReturn(documentFile);

            // call method to test
            Feed feed = new Feed(FEED_URL, null);
            LocalFeedUpdater.updateFeed(feed, context);
        }
```

damoasda replied:
Ok, I see what you mean. I extracted repetitive code into helper methods, for mocking external storage as mentioned above, and also the checks for the number of feeds and the number of items in the feed. The test methods look very simple now.

ByteHamster reviewed (commented):
Thanks for bringing roboelectric to my attention. It looks pretty promising - maybe we can even reduce the time needed for running some of the other existing tests by using roboelectric instead of a full emulator. I created a new issue for that: #4569

In order to run the tests on my PC, I needed to add the following dependency. Could you please add that one to your PR (if it does not break anything)?

    testImplementation 'javax.inject:javax.inject:1'

damoasda commented:
> In order to run the tests on my PC, I needed to add the following dependency. Could you please add that one to your PR (if it does not break anything)?
> 
> ```
> testImplementation 'javax.inject:javax.inject:1'
> ```

I added the dependency to the file `core/build.gradle`. For me `gradle test` works with and without this line. It doesn't break anything.

ByteHamster commented on core/src/test/java/androidx/documentfile/provider/AssetsDocumentFile.java:
```diff
@@ -0,0 +1,142 @@
+package androidx.documentfile.provider;
```
Could you please move this to our own namespace, eg. `de.danoeh.antennapod.core.storage`? The only reason why we have some classes in foreign packages is that we implemented workarounds that require access to package-private members.

damoasda replied:
That's exactly the reason why I had to use that package: The constructor of the super class `DocumentFile` is package-accessible only. If you prefer, I can change back that part of the test code and use Mockito instead.

ByteHamster replied:
👍 I did not notice that the constructor is protected when having a quick look at the code.

ByteHamster commented on core/src/test/java/androidx/documentfile/provider/AssetsDocumentFile.java:
```diff
@@ -0,0 +1,142 @@
+        if (fileName.endsWith(".mp3")) {
+            return "audio/mp3";
+        } else if (fileName.endsWith(".png")) {
+            return "image/png";
```
How about using `MimeTypeMap.getMimeTypeFromExtension`? That would work for more file types, just in case we used the class for more things later.

damoasda replied:
Good point. I made the code more generic by using the suggested method.

Note: When using Robolectric the class `MimeTypeMap` is replaced by the class `ShadowMimeTypeMap` that is empty by default. So I had to call the method `addExtensionMimeTypMapping()` to register the `mp3` file extension. I already had that line of code, removed it, and now it is back again. :grin:

ByteHamster replied:
Oh okay. As you can probably already tell, I have never worked with roboelectric before :) Your change looks good.

ByteHamster commented:
Awesome, thanks!

ByteHamster commented:
@damoasda 

CI recently started to fail regularly with this exception:

```
    java.lang.IllegalStateException: Illegal connection pointer 403. Current pointers for thread Thread[SDK 30 Main Thread,5,SDK 30] []
        at org.robolectric.shadows.ShadowSQLiteConnection$Connections.getConnection(ShadowSQLiteConnection.java:376)
        at org.robolectric.shadows.ShadowSQLiteConnection$Connections.prepareStatement(ShadowSQLiteConnection.java:435)
        at org.robolectric.shadows.ShadowSQLiteConnection.nativePrepareStatement(ShadowSQLiteConnection.java:91)
        at android.database.sqlite.SQLiteConnection.nativePrepareStatement(SQLiteConnection.java)
        at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:1045)
        at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:652)
        at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:590)
        at android.database.sqlite.SQLiteProgram.__constructor__(SQLiteProgram.java:61)
        at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java)
        at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java)
        at android.database.sqlite.SQLiteDatabase.delete(SQLiteDatabase.java:1727)
        at de.danoeh.antennapod.core.storage.PodDBAdapter.deleteDatabase(PodDBAdapter.java:381)
        at de.danoeh.antennapod.core.feed.LocalFeedUpdaterTest.setUp(LocalFeedUpdaterTest.java:73)
```

My guess would be that #4666 is responsible. You have read a bit more about that crash. Do you know what I did wrong in that PR?

damoasda commented:
@ByteHamster: The issue with Robolectric is that you have to avoid to reuse a database connection for several unit test threads. Last time I had to add the line
```
db = null;
```
in `PodDBAdapter.tearDownTests()` to make all tests running. Can you add that line again and try? (For that you will have to remove the `final` keyword from `db` and to initialize `db` in the `open()` method instead of the constructor, I think.)

BTW:
I tried to reproduce the problem but the unit tests for the master branch work fine on my computer. Also, I cannot see an error in CircleCI for PR #4666.

ByteHamster commented:
> Also, I cannot see an error in CircleCI for PR #4666.

I recently have to press the retry button a dozen times :) One example for CI failing in a recent PR: https://app.circleci.com/pipelines/github/AntennaPod/AntennaPod/2602/workflows/ae44de06-930a-455c-862c-e9f90e4483fa/jobs/1487

> you will have to remove the final keyword from db and to initialize db in the open() method

I thought that it is enough to set the instance to null. The tests then don't reuse the whole PodDBAdapter object (including DB). Apparently, it sometimes still fails and I don't understand why.

damoasda commented:
> I recently have to press the retry button a dozen times :)

Good hint, now I tried the tests several times - and I got the error, too :smile:
I will investigate to find a solution.

> I thought that it is enough to set the instance to null.

Sorry, you are right, your code looks semantically equivalent to the previous version.

ByteHamster commented:
> I will investigate to find a solution.

Thanks!

By the way, I have not forgotten your comment in #4661. I'm just pretty busy currently and don't have the patience to think about big decisions like that.

damoasda commented:
The reason for the occasional fails of `LocalFeedUpdaterTest` is probably the concurrency caused by `dbExec.submit()` calls in `DBWriter`. I found two different approaches that solved the problem on my computer (no fails in more than 20 test executions):

**1) Modify `DBWriter`**

Most methods with return type `Future<?>` are called without using the return value. After I changed the return type to `void` and removed the `dbExec.submit()` call the problem disappeared.

Drawback: Android doesn't like too much work handled in UI threads. This could lead to warnings in the app at runtime.

**2) Modify `LocalFeedUpdaterTest`**

Alternatively, I could solve the problem by adding the following lines to the method `LocalFeedUpdaterTest.setUp()`. It's a hack that overwrites the private field `DBWriter.dbExec` by a direct executor to avoid using different threads.

```
   public void setUp() throws Exception {
        // Don't fork processes in DBWriter
        ExecutorService directExecutorService = new AbstractExecutorService() {

            @Override
            public void execute(Runnable command) {
                command.run();
            }

            @Override
            public void shutdown() {
            }

            @Override
            public List<Runnable> shutdownNow() {
                return null;
            }

            @Override
            public boolean isShutdown() {
                return false;
            }

            @Override
            public boolean isTerminated() {
                return false;
            }

            @Override
            public boolean awaitTermination(long timeout, TimeUnit unit) {
                return false;
            }
        };
        Field field = DBWriter.class.getDeclaredField("dbExec");
        field.setAccessible(true);
        @SuppressWarnings("JavaReflectionMemberAccess")
        Field modifiersField = Field.class.getDeclaredField("modifiers");
        modifiersField.setAccessible(true);
        modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
        field.set(null, directExecutorService);

        ... (everything else)
```

Do you feel comfortable with one of the two approaches? If yes, I could provide an appropriate pull request.

> By the way, I have not forgotten your comment in #4661. I'm just pretty busy currently and don't have the patience to think about big decisions like that.

No stress, that topic hasn't been covered for years, so it won't run away. :smirk:

damoasda commented:
The "Illegal connection pointer" problem happened again in a PR. To continue the discussion and to solve the problem I created issue #4775.
